### ISSUE TITLE
[web] switch by runtimeType doesn't work for release build.

### ISSUE THREAD

**Issue by SunnyDayDev** [flutter/flutter#43154]
As mentioned in the title switch by runtimeType doesn't work for release build, but works fine for debug. Also it works fine on Android and iOS in both modes.

## Steps to Reproduce

if you run this code in debug mode you'll get the red screen, but in release mode you'll get the black screen.

```dart
import 'package:flutter/material.dart';

void main() => runApp(app());

Widget app() {
  Switcher switcher = Red();
  
  switch (switcher.runtimeType) {
    case Red: return Container(color: Colors.red,);
    default: return Container(color: Colors.black,);
  }
}

abstract class Switcher {}

class Red implements Switcher {}
```

**Target Platform: Web**
**Target OS version/browser: macOS 10.15.1 / Safari 13.03 /Chrome  77.0.3865.120**

## Logs

flutter doctor
```
[✓] Flutter (Channel dev, v1.10.14, on Mac OS X 10.15.1 19B77a, locale ru-RU)
    • Flutter version 1.10.14 at /Users/sunny/Library/Flutter
    • Framework revision 1946fc4da0 (13 days ago), 2019-10-07 15:23:31 -0700
    • Engine revision 1d62160fdb
    • Dart version 2.6.0 (build 2.6.0-dev.1.0 d6c6d12ebf)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/sunny/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/sunny/Library/Android/sdk
    • Java binary at: /Users/sunny/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/191.5900203/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.1, Build version 11A1027
    • CocoaPods version 1.8.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Users/sunny/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/191.5900203/Android Studio.app/Contents
    • Flutter plugin version 40.2.2
    • Dart plugin version 191.8580
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Ultimate Edition (version 2019.2.3)
    • IntelliJ at /Users/sunny/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (2 available)
    • Chrome          • chrome          • web-javascript • Google Chrome 77.0.3865.120
    • Headless Server • headless-server • web-javascript • Flutter Tools

! Doctor found issues in 1 category.
```

**Comment by vsmenon** [comment]
@sigmundch

**Comment by sigmundch** [comment]
Thanks for filing the issue and the small example!

Turns out this is a well known limitation of the release dart2js compiler (https://github.com/dart-lang/sdk/issues/17207). We allow using type literals, but we don't canonicalize types so they are not available to be used in switch statements. At the moment, the best workaround for now is to use if-then-else instead of switch blocks.

The good news is that we are very close to fixing this. As part of the work to support NNBD we have made a large effort to change the runtime type representation. The new representation will support this use case. The new representation is not complete though, so it will be a little more time before we can close the bug.

**Comment by vsmenon** [comment]
@sigmundch - flutter is on the new rti now, right?

**Comment by sigmundch** [comment]
it's on by default everywhere, but it depends on which version of Dart is being used by flutter. This was enabled on Dart 2.8.0-dev.1.0 for the first time. Assuming flutter is using a later version (or master if we are autorolling), this issue should be now resolved.

**Comment by SunnyDayDev** [comment]
I confirm that it works fine on the latest Flutter which I use.
`
flutter --version
Flutter 1.14.6 • channel dev • https://github.com/flutter/flutter.git
Framework • revision fabeb2a16f (8 days ago) • 2020-01-28 07:56:51 -0800
Engine • revision c4229bfbba
Tools • Dart 2.8.0 (build 2.8.0-dev.5.0 fc3af737c7)
`

**Comment by jonahwilliams** [comment]
🎊 Yay! Closing as fixed then

**Comment by deisold** [comment]
Hi, I'm still having this issue with current master:

[✓] Flutter (Channel master, v1.15.4-pre.51, on Mac OS X 10.15.2 19C57, locale de-DE)
    • Flutter version 1.15.4-pre.51 at /programming/flutter
    • Framework revision dc636069fc (13 minutes ago), 2020-02-12 22:22:49 -0800
    • Engine revision e0ebaea590
    • Dart version 2.8.0 (build 2.8.0-dev.9.0 e4c39721c4)

@jonahwilliams: seems to be a dart issue. I also changed back to flutter beta, but the same

**Comment by iapicca** [comment]
Could everyone who still has this problem please file a new issue with the exact descriptions what happens, logs and the output of 'flutter doctor -v'  please.
All system setups can be slightly different so it's always better to open new issues and reference related issues.

**Comment by shinriyo** [comment]
```
flutter build web
cd build/web/
firebase serve
```

nothing is shown...

but, debug build works.

**Comment by ductranit** [comment]
This issue still happens on flutter web release on `master` channel
```
I tried: CrashReport.runtimeType.toString();
On web debug: CrashReport
On web release: minified:Dr
```
It looks like the dart2js minified the type name on release build

flutter doctor:
```
[✓] Flutter (Channel master, 1.22.0-2.0.pre.105, on Mac OS X 10.15.4 19E287, locale en-US)
    • Flutter version 1.22.0-2.0.pre.105 at /Users/root/Documents/Soft/flutter
    • Framework revision cad4d1333e (8 hours ago), 2020-08-27 17:48:05 -0700
    • Engine revision 77dd1c05b9
    • Dart version 2.10.0 (build 2.10.0-67.0.dev)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/root/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/root/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/root/Library/Android/sdk
    • Java binary at: /Applications/Android Studio 3.5
      Preview.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.
[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.8.4
[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
[!] Android Studio
    • Android Studio at /Applications/Android Studio 4.0 Preview.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio 3.5 Preview.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2
[✓] Connected device (3 available)
    • macOS (desktop)  • macos      • darwin-x64     • Mac OS X 10.15.4 19E287
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 84.0.4147.135
! Doctor found issues in 2 categories.
```

**Comment by sigmundch** [comment]
@deisold - I'm sorry I missed your original post, @ductranit - thanks for following up

I know it's not the answer you were hoping for, but this is one behavior of .runtimeType that is by design.

The Dart language specification purposely allows the `runtimeType.toString` to be changed under different circumstances. The dart2js compiler changes it when minifying applications to remove all references to the original names, this helps both with code-size and with sensitive data (many teams find that the names used in their code are too sensitive if the app is not open source, and prefer to keep those private).

The main purpose for `.runtimeType.toString` is that when errors occur (e.g. exceptions are thrown), they may contain a runtime type in the error message. These errors can be logged to a server where you can use data in the source-map file to lookup the original name and triage a production failure. Beyond that, our general recommendation is that it should not be used for almost anything else: no control-flow logic (we recommend subtype `is` checks instead), no comparing it against expected names, no displaying it visibly on the UI. See also https://github.com/dart-lang/sdk/issues/35052#issuecomment-621644698 .

Sorry I don't have better news for you, but hopefully this additional context is helpful.

**Comment by ductranit** [comment]
@sigmundch thank for your information, that makes sense. I'll try another way
